bot_settings: Add status based filter in SETTINGS/BOTS.

shuklamaneesh23 opened the pull request:
**Commit 1:**

- Added a status-based filter to the Bots list in the SETTINGS/BOTS section.
- Users can now filter bots by their status (Active or Deactivated) using a dropdown menu located to the left of the "Filter bots" field.
- The default selection for the filter is set to "Active" when the panel is opened.

Fixes part of #31156.

<!-- Describe your pull request here.-->

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Screenshots and screen captures:**

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [ ] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [x] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

shuklamaneesh23 commented:
@zulipbot add "buddy review"

@kuv2707 , could you please take a look at it?

kuv2707 commented on web/src/settings_users.ts:
```diff
@@ -264,8 +283,13 @@ function status_selected_handler(
-    add_value_to_filters(bots_section, "status_code", status_code);
-
+    if (widget.widget_name === all_bots_section.dropdown_widget_name) {
+        // all_bots_section.list_widget?.set_dropdown_value(status_code);
```
This could be removed.

kuv2707 commented on web/src/settings_users.ts:
```diff
@@ -264,8 +283,13 @@ function status_selected_handler(
+        // all_bots_section.list_widget?.set_dropdown_value(status_code);
+        add_value_to_filters(all_bots_section, "status_code", status_code);
+    } else if (widget.widget_name === your_bots_section.dropdown_widget_name) {
+        // your_bots_section.list_widget?.set_dropdown_value(status_code);
```
This too.

kuv2707 commented on web/src/settings_users.ts:
```diff
@@ -416,6 +440,7 @@ type BotInfo = {
     role: number;
     status_code: number;
     is_active: boolean;
+    can_modify_bot: boolean;
```
I don't understand why this needs to be on `BotInfo`. It doesn't seem to be read anywhere, and seems to converge with the existing `can_modify`

kuv2707 commented on web/src/settings_users.ts:
```diff
@@ -536,50 +569,96 @@ function set_text_search_value($table: JQuery, value: string): void {
+    $all_bots_table.show();
+}
+
+function your_bots_create_table(): void {
```
This could be better named something like `create_table_your_bots`.

shuklamaneesh23 replied:
Changed to `create_your_bots_table` and `create_all_bots_table`.

kuv2707 commented on web/templates/settings/bot_list_admin.hbs:
```diff
@@ -11,40 +11,17 @@
+      show_botserverrc=false
+      }}
+    {{> bot_list
+      section="your"
```
This might also be better named `your_bots`.

kuv2707 commented on web/templates/settings/bot_list_admin.hbs:
```diff
@@ -22,6 +22,7 @@
       dropdown_widget_name=your_bots_list_dropdown_widget_name
-      show_botserverrc=false
+      show_botserverrc=true
+      botserverrc=botserverrc
```
Does this change belong to this commit only?

shuklamaneesh23 replied:
Yes, this is intentional.
I've added `show_botserverrc` in the hbs partial in the previous commit and logic to render that in this commit for the ease of review.

kuv2707 commented:
<img width="584" alt="image" src="https://github.com/user-attachments/assets/4f0f2d01-f986-4594-830c-5508e5d692af" />

I'm not the person to suggest UI changes but it looks too sparse and scattered. Changes would better be suggested by some other maintainer having expertise in that area.

alya commented:
I haven't tried to comprehensively test all the functionality, but this is looking great! I didn't see any problems as I was poking around.

sujalshah-bit commented:
@zulipbot add "maintainer review"

sujalshah-bit commented:
my review is ongoing. I am adding a maintainer label so that PR's progress should not be delayed.

sahil839 reviewed (commented):
Posted some comments.

shuklamaneesh23 commented:
Hi @sahil839 , thanks for such a detailed review.
I've addressed all the review and posted my doubts whatever i had in my mind.
Please have a look whenever you have a moment.

sahil839 commented on web/src/settings_users.ts:
```diff
@@ -812,28 +898,32 @@ function deactivated_handle_events(): void {
-    active_section.handle_events();
-    deactivated_section.handle_events();
+    active_handle_events();
+    deactivated_handle_events();
```
I may not have read that before. I am not sure if that has an advantage and if any it should be done in a separate commit. We can skip it for now and see later if we need that change.

sahil839 reviewed (commented):
Posted some comments.

I think we might need some refactoring of `settings_users.ts` code and also maybe code in `settings_panel_menu.ts`, especially the code for handling tabs, but I do not have a clear plan in my mind right now and would require some trial and error, so we can do that later.

shuklamaneesh23 commented on web/templates/settings/bot_list.hbs:
```diff
@@ -0,0 +1,29 @@
+                <th class="actions">{{t "Actions" }}</th>
+            </thead>
+            <tbody id="admin-{{section}}-bots-table"
+              data-empty="{{#if (eq section 'all')}}{{t 'There is no active bot.' }}{{else}}{{t 'You have no active bots.' }}{{/if}}"
```
Yes, we can do that, but since the "Active" case is the default, I thought it would make sense to have a different string here. This is the state users will see by default when they open the "Bots" section without applying any filters, so showing "No bots match your current filter." might feel confusing to them.

sahil839 commented on web/src/hashchange.ts:
```diff
@@ -89,6 +89,9 @@ function handle_invalid_section_url(section: "bots" | "users", settings_tab: str
 function get_settings_tab(section: string): string | undefined {
     if (section === "users" || section === "bots") {
         const current_tab = hash_parser.get_current_nth_hash_section(2);
+        if (section === "bots" && window.location.hash.includes("your-bots")) {
```
Yes, but do we need this at all? What url will this handle?

shuklamaneesh23 commented on web/src/settings_users.ts:
```diff
@@ -751,10 +830,16 @@ export function redraw_active_users_list(): void {
+export function rerender_bot_section(bot_user_id: number): void {
+    redraw_all_bots_list();
+    redraw_your_bots_list();
+    update_bot_data(bot_user_id);
```
Yes, this is needed for updating the text `Download config of all active outgoing webhook bots in Zulip Botserver format.`.

sahil839 commented:
I have done some updates. Have mostly added some refactoring commits and some renames to variables and class and IDs of elements. Also removed unnecessary or unused code at a couple of places.

Not marking it ready for review, as I still need to test it manually once more and probably some changes to CSS and commit order as well.

sahil839 commented:
Updated the PR. Made some changes in refactoring commits and added some missing live update code.

@alya had a couple of questions -
- We currently allow regenerating the API key from "Manage bot" modal and the current behavior is the API key is regenerated on clicking the button. Is it fine to do so when other actions require clicking on "Save changes" button? I think it is fine for now given how the API is designed currently.
- Do we want to remove the bot row when deactivating the bot immediately and add it immediately while activating it? Like for users, we do not remove or add the row immediately, but there is we use tabs for active and deactivated users and here we have dropdown

alya commented:
Good questions!

1. I think the ideal behavior would require confirmation to change the API key, but probably let's leave that for a follow-up.
2. I like having the row there (as it is now).

alya commented:
Oh, I see. That seems good, but not a blocker for merging this PR if it's non-trivial -- if I'm thinking about it right, it wouldn't come up too much.

sahil839 commented:
Yes, it will only come up when deactivating or reactivating a bot. And it is somehwat non-trivial to keep the bot row in case it is visible but to add it if it is not visible and this PR is already large, so will leave it for now.

The last push was for some CSS changes which does not have any visual effect.

sahil839 commented:
@timabbott this is ready for your review. I think we can do some more refactors like renaming classes in a better way and also the refactor discussed in https://chat.zulip.org/#narrow/channel/6-frontend/topic/splitting.20settings_users.2Ets/with/2266495.

alya commented:
> Another quirk, which seems like it should get its own issue, is the bot cards cut off the name of the bot owner very early:

I think it's fine -- @timabbott probably had it zoomed in. Even at max font size:

<img width="346" height="225" alt="Screenshot 2025-11-11 at 22 39 02" src="https://github.com/user-attachments/assets/6bc49ef6-dcd8-4a32-b402-0282a4ba7933" />
